File: src/balancer.c
/* balancer.c
 * Outgoing interface abstraction and load balancing algorithm
 * 
 * Copyright 2015-2018 Akash Rawal
 * This file is part of dispatch_ng.
 * 
 * dispatch_ng is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * dispatch_ng is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with dispatch_ng.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "incl.h"

//Errors
define_static_error(balancer_error_no_iface, "No interfaces are available");

//Interface data structure
struct _Interface
{
	int index;
	int metric;
	int use_count;
	HostAddress addr;
};
NetworkType types = 0;

//Min-heap data structure
typedef struct
{
	Interface **data;
	size_t len, alloc_len;
} Heap;

Heap ip4[1] = {0}, ip6[1] = {0};

static double value(Heap *heap, int idx)
{
	return ((double) heap->data[idx]->use_count / heap->data[idx]->metric);
}

static void assign(Heap *heap, int idx, Interface *value)
{
	heap->data[idx] = value;
	if (value)
		value->index = idx;
}

#if 0
static void assert_heap(Heap *heap)
{
	int i, parent;

	for (i = 1; i < heap->len; i++)
	{
		parent = (i - 1) / 2;
		abort_if_fail(value(heap, parent) <= value(heap, i),
				"Assertion failure");
	}
}
#endif

static void shift_up(Heap *heap, int idx)
{
	Interface *tmp = heap->data[idx];
	double tmp_value = value(heap, idx);
	while (idx > 0)
	{
		int parent = (idx - 1) / 2;
		if (value(heap, parent) <= tmp_value)
			break;
		assign(heap, idx, heap->data[parent]);
		idx = parent;
	}
	assign(heap, idx, tmp);

	//assert_heap(heap);
}

static void shift_down(Heap *heap, int idx)
{
	Interface *tmp = heap->data[idx];
	double tmp_value = value(heap, idx);

	while (idx < heap->len)
	{
		int left = idx * 2 + 1;
		int right = idx * 2 + 2;
		int sel = -1;

		if (left < heap->len)
			sel = left;

		if (right < heap->len)
		{
			if (value(heap, right) < value(heap, left))
				sel = right;
		}

		if (sel >= 0)
		{
			if (value(heap, sel) < tmp_value)
			{
				assign(heap, idx, heap->data[sel]);
				idx = sel;
				continue;
			}
		}
		break;
	}

	abort_if_fail(idx < heap->len, "Assertion failure");

	assign(heap, idx, tmp);

	//assert_heap(heap);
}

static void heap_insert(Heap *heap, Interface *iface)
{
	if (heap->alloc_len == 0)
	{
		heap->alloc_len = 8;
		heap->data = fs_malloc(sizeof(void *) * heap->alloc_len);
	}
	else if (heap->alloc_len >= heap->len)
	{
		heap->alloc_len *= 2;
		heap->data = fs_realloc(heap->data, sizeof(void *) * heap->alloc_len);
	}
	heap->len++;
	assign(heap, heap->len - 1, iface);
	shift_up(heap, heap->len - 1);
}

static void heap_delete(Heap *heap, Interface *iface)
{
	abort_if_fail(heap->len && iface->index >= 0 && iface->index < heap->len, 
			"Assertion failure");

	int idx = iface->index;

	iface->index = -1;
	heap->len--;
	if (idx < heap->len - 1)
	{
		heap->data[idx] = heap->data[heap->len - 1];
		shift_down(heap, idx);	
	}
}

static Heap *select_heap(Interface *iface)
{
	if (iface->addr.type == NETWORK_INET)
		return ip4;
	else
		return ip6;
}

void interface_close(Interface *iface)
{
	iface->use_count--;
	shift_up(select_heap(iface), iface->index);
}

HostAddress interface_get_addr(Interface *iface)
{
	return iface->addr;
}

Interface *balancer_add(HostAddress addr, int metric)
{
	Interface *iface;

	iface = (Interface *) fs_malloc(sizeof(Interface));
	
	iface->index = -1;
	iface->addr = addr;
	iface->metric = metric;
	if (iface->metric < 0)
		iface->metric = 1;
	iface->use_count = 0;
	
	heap_insert(select_heap(iface), iface);
	types |= addr.type;
	
	return iface;
}

Interface *balancer_add_from_string(const char *addr_with_metric)
{
	char *addr_str, *metric_str;
	HostAddress addr;
	int metric;
	Status s;

	addr_str = split_string(addr_with_metric, '@', &metric_str);

	if (addr_str)
	{
		char *endptr;
		s = host_address_from_str(addr_str, &addr);	
		metric = strtol(metric_str, &endptr, 10);
		if (*endptr)
			s = STATUS_FAILURE;
		free(addr_str);
	}
	else
	{
		s = host_address_from_str(addr_with_metric, &addr);	
		metric = -1;
	}

	abort_if_fail(s == STATUS_SUCCESS,
			"Failed to parse address %s", addr_with_metric);

	return balancer_add(addr, metric);
}

/* Verify that all addressses can be bound to. 
 * (Windows has a wierd problem that some IP addresses magically fail to bind.
 * No explanation yet. Hence exists this workaround to make our lives a 
 * bit easier.)
 */
void balancer_verify()
{
	int i, j;
	int n_fails = 0;
	Heap *heaps[2] = { ip4, ip6 };
	for (i = 0; i < 2; i++)
	{
		if (heaps[i]->len == 0)
			continue;

		//Copy
		Interface **ifaces = fs_malloc(sizeof(void *) * heaps[i]->len);
		for (j = 0; j < heaps[i]->len; j++)	
			ifaces[j] = heaps[i]->data[j];

		for (j = 0; j < heaps[i]->len; j++)	
		{
			SocketHandle hd;
			SocketAddress addr;
			addr.host = ifaces[j]->addr;
			addr.port = 0;
			const Error *e = socket_handle_create_bound(addr, &hd);
			if (e)
			{
				char str[ADDRESS_MAX_LEN];
				host_address_to_str(addr.host, str);
				fprintf(stderr, 
						"Warning: Address %s is unusable, removing (%s)\n",
							str, error_desc(e));
				error_handle(e);
				heap_delete(heaps[i], ifaces[j]);
				n_fails++;
			}
		}
		free(ifaces);
	}
	if (n_fails > 0)
	{
		fprintf(stderr, "Warning: %d addresses are not usable\n", n_fails);
	}
	if (ip4->len == 0 && ip6->len == 0)
	{
		fprintf(stderr, "Error: No addresses remaining!\n");
		exit(2);
	}
}

define_static_error(balancer_struct_no_iface,
		"No suitable interface available");

const Error *balancer_open_iface(NetworkType types,
		Interface **iface_out, SocketHandle *hd_out)
{
	Heap *heaps[2] = { NULL, NULL };
	Interface *selected = NULL;
	Heap *selected_heap = NULL;
	SocketHandle hd;
	const Error *e = NULL;
	int i;

	if (types & NETWORK_INET)
		heaps[0] = ip4;
	if (types & NETWORK_INET6)
		heaps[1] = ip6;

	for (i = 0; i < 2; i++)
	{
		if (!heaps[i])
			continue;

		if (heaps[i]->len == 0)
			continue;


		if (selected_heap)
		{
			if (value(heaps[i], 0) >= value(selected_heap, 0))
				continue;
		}

		selected_heap = heaps[i];
	}

	if (! selected_heap)
		return balancer_error_no_iface_instance;

	selected = selected_heap->data[0];

	{
		SocketAddress addr;
		addr.host = selected->addr;
		addr.port = 0;
		e = socket_handle_create_bound(addr, &hd);
	}
	if (e)
		return e;

	selected->use_count++;
	shift_down(selected_heap, selected->index);
	*iface_out = selected;
	*hd_out = hd;
	return NULL;
}

NetworkType balancer_get_available_types()
{
	return types;
}

void balancer_shutdown()
{
	int i, j;
	Heap *heaps[2] = { ip4, ip6 };

	for (i = 0; i < 2; i++)
	{
		if (! heaps[i]->data)
			continue;

		for (j = 0; j < heaps[i]->len; j++)
			free(heaps[i]->data[j]);
		free(heaps[i]->data);

		memset(heaps[i], 0, sizeof(Heap));
	}

	types = 0;
}

File: src/balancer.rs
/* balancer.rs
 * Load balancer algorithm, selects the outgoing IP address with minimum 
 * use count.
 * 
 * Copyright 2023 Akash Rawal
 * This file is part of dispatch_ng.
 * 
 * dispatch_ng is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * dispatch_ng is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with dispatch_ng.  If not, see <http://www.gnu.org/licenses/>.
 */

use std::hash::Hash;

use std::sync::Arc;
use std::sync::Mutex;

use std::net::{ IpAddr, Ipv4Addr, Ipv6Addr };

use crate::ord::{ OrdWrapper, EasyOrd };
use crate::balance_heap::BalanceHeap;


//Metric state (stores use count and provides comparison function)
#[derive(Clone)]
pub struct MetricState {
    use_count : usize,
    metric : usize,
}

impl EasyOrd for MetricState {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        let self_ord = self.use_count / self.metric;
        let other_ord = other.use_count / other.metric;
        other_ord.cmp(&self_ord) //< Reverse order
    }
}


//Balancer state
pub struct BalancerState {
    v4_heap : BalanceHeap<Ipv4Addr, OrdWrapper<MetricState>>, 
    v6_heap : BalanceHeap<Ipv6Addr, OrdWrapper<MetricState>>, 
}

impl<T> FromIterator<T> for BalancerState 
where Self : Extend<T> {
    fn from_iter<U: IntoIterator<Item = T>>(iter: U) -> Self {
        let mut res = Self {
            v4_heap : BalanceHeap::new(),
            v6_heap : BalanceHeap::new(),
        };
        res.extend(iter);
        res
    }
}

impl Extend<(IpAddr, usize)> for BalancerState {
    fn extend<T: IntoIterator<Item = (IpAddr, usize)>>(&mut self, iter: T) {
        for (addr, metric) in iter {
            let order_value = OrdWrapper(MetricState {
                use_count : 0,
                metric,
            });
            match addr {
                IpAddr::V4(v4addr) => self.v4_heap.update(v4addr, order_value),
                IpAddr::V6(v6addr) => self.v6_heap.update(v6addr, order_value),
            } 
        }
    }
}

impl Extend<IpAddr> for BalancerState {
    fn extend<T: IntoIterator<Item = IpAddr>>(&mut self, iter: T) {
        self.extend(iter.into_iter().map(|addr| (addr, 1)))
    }
}

impl BalancerState {
    fn chg_use_count(&mut self, addr : IpAddr, delta : isize) {
        fn update_heap<T : Hash + Eq + Clone>(
            heap : &mut BalanceHeap<T, OrdWrapper<MetricState>>, 
            value : T,
            delta : isize
        ) {
            let OrdWrapper(MetricState { use_count, metric }) 
                = heap.get(&value).unwrap().clone();
            heap.update(value, OrdWrapper(MetricState { 
                use_count : ((use_count as isize) + delta) as usize, 
                metric,
            }));
        }

        match addr {
            IpAddr::V4(addr) => update_heap(&mut self.v4_heap, addr, delta),
            IpAddr::V6(addr) => update_heap(&mut self.v6_heap, addr, delta),
        }
    }
}

//Balancer
#[derive(Clone)]
pub struct Balancer(Arc<Mutex<BalancerState>>);

impl Balancer {
    pub fn new(state : BalancerState) -> Self {
        Self(Arc::new(Mutex::new(state)))
    }
}

pub struct IpGuard {
    balancer : Balancer,
    addr : IpAddr,
}

impl IpGuard {
    pub fn get_addr(&self) -> IpAddr {
        self.addr
    }
}


impl Drop for IpGuard {
    fn drop(&mut self) {
        let mut lock = self.balancer.0.lock().unwrap();
        lock.chg_use_count(self.addr, -1);
    }
}

impl Balancer {
    pub fn take(&self, take_v4 : bool, take_v6 : bool) -> Option<IpGuard> {
        let addr = {
            let mut lock = self.0.lock().unwrap();
            let mut list = Vec::<(IpAddr, OrdWrapper<MetricState>)>::new();
            if take_v4 {
                if let Some((addr, met)) = lock.v4_heap.peek() {
                    list.push((IpAddr::V4(addr), met));
                }
            }
            if take_v6 {
                if let Some((addr, met)) = lock.v6_heap.peek() {
                    list.push((IpAddr::V6(addr), met));
                }
            }
            list.sort_by(|a, b| a.1.cmp(&b.1));

            let addr = list.last()?.0; 
            lock.chg_use_count(addr, 1);

            addr
        };

        Some(IpGuard {
            balancer : self.clone(),
            addr
        })
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test() {
        let addr1 = IpAddr::from([172, 16, 0, 1]);
        let addr2 = IpAddr::from([172, 16, 0, 2]);

        let balancer = Balancer::new([ (addr1, 1), (addr2, 2) ]
                                     .into_iter().collect());

        let res1 = balancer.take(true, true);
        let res2 = balancer.take(true, true);
        let res3 = balancer.take(true, true);

        let vec : Vec<IpAddr> = res1.into_iter().chain(res2).chain(res3)
            .map(|guard| guard.addr)
            .collect();

        assert_eq!(vec.iter().filter(|x| **x == addr1).count(), 1);
        assert_eq!(vec.iter().filter(|x| **x == addr2).count(), 2);
    }
}
